File: backend/app/models/product.py
from pydantic import BaseModel, Field
from typing import Dict, List, Optional

from app.models.body import UserBodyMeasurements


class ProductUrlInput(BaseModel):
    url: str = Field(..., min_length=8)


class SizeMeasurement(BaseModel):
    size_label: str
    chest_cm: Optional[float] = None
    waist_cm: Optional[float] = None
    hip_cm: Optional[float] = None
    length_cm: Optional[float] = None
    shoulder_cm: Optional[float] = None
    sleeve_cm: Optional[float] = None
    biceps_cm: Optional[float] = None
    top_length_cm: Optional[float] = None
    inseam_cm: Optional[float] = None
    thigh_cm: Optional[float] = None
    rise_cm: Optional[float] = None
    wrist_cm: Optional[float] = None
    garment_category: Optional[str] = None
    stretch_level: Optional[str] = None
    additional_measurements: Dict[str, float] = Field(default_factory=dict)
    is_estimated: bool = False
    estimated_from_size: Optional[str] = None
    confidence: Optional[float] = None


class GarmentMeasurements(SizeMeasurement):
    pass


class FabricAnalysis(BaseModel):
    raw_text: Optional[str] = None
    stretch_factor: float = 0.0
    shrink_risk: float = 0.0
    drape_factor: float = 0.5
    detected_fibers: List[str] = Field(default_factory=list)
    warnings: List[str] = Field(default_factory=list)


class ProductScrapeResult(BaseModel):
    source_url: str
    title: str
    image_url: Optional[str] = None
    currency: Optional[str] = None
    price: Optional[str] = None
    raw_size_text: Optional[str] = None
    normalized_sizes: List[SizeMeasurement] = Field(default_factory=list)
    fabric_composition_text: Optional[str] = None
    fabric_analysis: Optional[FabricAnalysis] = None
    confidence_score: Optional[float] = None
    extraction_method: Optional[str] = None
    fallback_reason: Optional[str] = None
    blocked_by_antibot: bool = False


class GarmentUploadResult(BaseModel):
    filename: str
    content_type: str
    original_path: str
    processed_path: Optional[str] = None
    original_url: Optional[str] = None
    processed_url: Optional[str] = None
    background_removed: bool = False
    message: str


class FitCheckInput(BaseModel):
    user_chest_cm: float = Field(..., ge=40, le=220)
    user_waist_cm: float = Field(..., ge=35, le=220)
    user_hip_cm: float = Field(..., ge=40, le=240)
    user_length_cm: Optional[float] = Field(default=None, ge=20, le=180)
    user_sleeve_cm: Optional[float] = Field(default=None, ge=20, le=120)
    user_biceps_cm: Optional[float] = Field(default=None, ge=15, le=90)
    user_top_length_cm: Optional[float] = Field(default=None, ge=20, le=150)
    user_inseam_cm: Optional[float] = Field(default=None, ge=35, le=140)
    user_thigh_cm: Optional[float] = Field(default=None, ge=25, le=120)
    user_shoulder_cm: Optional[float] = Field(default=None, ge=25, le=90)
    user_rise_cm: Optional[float] = Field(default=None, ge=10, le=70)
    user_wrist_cm: Optional[float] = Field(default=None, ge=10, le=50)
    user_measurements: Optional[UserBodyMeasurements] = None
    garment_size: SizeMeasurement
    candidate_sizes: List[SizeMeasurement] = Field(default_factory=list)
    garment_category: Optional[str] = None
    stretch_level: Optional[str] = None
    fabric_analysis: Optional[FabricAnalysis] = None
    user_ease_modifier: float = Field(default=0.0, ge=-6, le=6)


class FitZone(BaseModel):
    zone: str
    difference_cm: Optional[float]
    status: str
    color: str
    message: str
    body_cm: Optional[float] = None
    garment_cm: Optional[float] = None
    ease_allowance_cm: Optional[float] = None
    pressure_score: Optional[float] = None
    fabric_warning: Optional[str] = None


class FitSizeOption(BaseModel):
    size_label: str
    score: float
    zones: List[FitZone]
    summary: str
    is_estimated: bool = False
    is_best_match: bool = False


class FitCheckResult(BaseModel):
    zones: List[FitZone]
    summary: str
    best_size_label: Optional[str] = None
    selected_size_label: Optional[str] = None
    size_options: List[FitSizeOption] = Field(default_factory=list)
    fabric_warnings: List[str] = Field(default_factory=list)
    cache_key: Optional[str] = None
    cache_hit: bool = False


class FitFeedbackInput(BaseModel):
    user_key: str = Field(default="local-device", min_length=3)
    predicted_status: str
    reported_status: str
    zone: Optional[str] = None


class FitFeedbackResult(BaseModel):
    user_key: str
    user_ease_modifier: float
    message: str

File: frontend/src/types/product.ts
export type ProductUrlInput = {
  url: string;
};

export type SizeMeasurement = {
  size_label: string;
  chest_cm?: number | null;
  waist_cm?: number | null;
  hip_cm?: number | null;
  length_cm?: number | null;
  shoulder_cm?: number | null;
  sleeve_cm?: number | null;
  biceps_cm?: number | null;
  top_length_cm?: number | null;
  inseam_cm?: number | null;
  thigh_cm?: number | null;
  rise_cm?: number | null;
  wrist_cm?: number | null;
  garment_category?: string | null;
  stretch_level?: string | null;
  additional_measurements?: Record<string, number>;
  is_estimated?: boolean;
  estimated_from_size?: string | null;
  confidence?: number | null;
};

export type FabricAnalysis = {
  raw_text?: string | null;
  stretch_factor: number;
  shrink_risk: number;
  drape_factor: number;
  detected_fibers: string[];
  warnings: string[];
};

export type ProductScrapeResult = {
  source_url: string;
  title: string;
  image_url?: string | null;
  currency?: string | null;
  price?: string | null;
  raw_size_text?: string | null;
  normalized_sizes: SizeMeasurement[];
  fabric_composition_text?: string | null;
  fabric_analysis?: FabricAnalysis | null;
  confidence_score?: number | null;
  extraction_method?: string | null;
  fallback_reason?: string | null;
  blocked_by_antibot?: boolean;
};

export type GarmentUploadResult = {
  filename: string;
  content_type: string;
  original_path: string;
  processed_path?: string | null;
  original_url?: string | null;
  processed_url?: string | null;
  background_removed: boolean;
  message: string;
};

export type FitCheckInput = {
  user_chest_cm: number;
  user_waist_cm: number;
  user_hip_cm: number;
  user_length_cm?: number | null;
  user_sleeve_cm?: number | null;
  user_biceps_cm?: number | null;
  user_top_length_cm?: number | null;
  user_inseam_cm?: number | null;
  user_thigh_cm?: number | null;
  user_shoulder_cm?: number | null;
  user_rise_cm?: number | null;
  user_wrist_cm?: number | null;
  garment_size: SizeMeasurement;
  candidate_sizes?: SizeMeasurement[];
  garment_category?: string | null;
  stretch_level?: string | null;
  fabric_analysis?: FabricAnalysis | null;
  user_ease_modifier?: number;
};

export type FitZone = {
  zone: string;
  difference_cm?: number | null;
  status:
    | "apertado"
    | "justo"
    | "folgado"
    | "sem_informacao"
    | "too_small"
    | "tight"
    | "balanced"
    | "loose"
    | "unknown"
    | string;
  color: "red" | "yellow" | "green" | "blue" | "gray" | string;
  message: string;
  body_cm?: number | null;
  garment_cm?: number | null;
  ease_allowance_cm?: number | null;
  pressure_score?: number | null;
  fabric_warning?: string | null;
};

export type FitSizeOption = {
  size_label: string;
  score: number;
  zones: FitZone[];
  summary: string;
  is_estimated: boolean;
  is_best_match: boolean;
};

export type FitCheckResult = {
  zones: FitZone[];
  summary: string;
  best_size_label?: string | null;
  selected_size_label?: string | null;
  size_options?: FitSizeOption[];
  fabric_warnings?: string[];
  cache_key?: string | null;
  cache_hit?: boolean;
};

export type FitFeedbackInput = {
  user_key?: string;
  predicted_status: string;
  reported_status: string;
  zone?: string | null;
};

export type FitFeedbackResult = {
  user_key: string;
  user_ease_modifier: number;
  message: string;
};
